### ISSUE TITLE
Adding display-referred color spaces

### ISSUE THREAD

**Issue by doug-walker** [AcademySoftwareFoundation/OpenColorIO#926]
This proposal is to add a fundamental new capability to OCIO.  It goes beyond a simple new feature and reflects an enhancement to the underlying architecture.  As such this is a lengthy proposal which begins with some background information to define the problem and then proceeds with a specific proposal.  

The issue is divided into sections, so you may skip down to the config and API proposal sections if you are so inclined.  

Please note that the proposal presented below would be an optional feature and config authors would not be required to use it or to otherwise modify their configs if they don't find it useful.

**BACKGROUND**

ISO standard 22028-1 introduced the notion of "image state" and scene-referred and output-referred (or, more specifically, display-referred) color encodings.  That document forms the foundation for many other color-management-related standards that are essential to the VFX industry including those for ACES (i.e., the SMPTE ST 2065 series of standards) and HDR video (i.e., ITU-R BT.2100, BT.2390, etc.).

One of the core ideas of these standards is that there is a special type of transform needed to convert from a scene-referred encoding to a display-referred encoding.  In ISO 22028-1 and ACES, this is called a "rendering transform" (which unfortunately overlaps with the use of "rendering" elsewhere in VFX).  In the ITU standards it is called the "opto-optical transfer function" or OOTF.

Many VFX artists are missing this key conceptual piece of information and incorrectly assume that there is no fundamental difference between scene-referred and display-referred encodings.  Much of the information on the web about the so called "linear workflow" for CG is wrong due to this misunderstanding.

OCIO was quite forward-thinking in terms of introducing the notion of a "view" that is a separate thing from a "display".  Hopefully this is gradually helping artists realize that the proper conversion from a scene-referred color space to a display is generally not simply a gamma correction (or inverse EOTF, to use the ITU term of art).  

However, currently in OCIO the view/display terminology is basically just a UI-related naming feature and the underlying architecture does not reflect these concepts.  For example, there is no notion of a display ColorSpace.  There is also no notion of a "view transform" that connects a scene-referred and display-referred color space.  The display/view section of the config is simply a way of stating that certain of the color spaces are actually transforms to a display space that incorporate a specific view.  For example, in the spi-vfx config there is a "Film" view for a DCIP3 and an sRGB display.  These point to transforms that convert from the scene-referred reference space all the way to a display (p3dci8 and srgb8).  This is inconvenient for several reasons.

First, it means that there are no color spaces for displays.  In other words, there is no color space simply for "sRGB" or "DCIP3" that does not also have a view transform (a film emulation in the example) baked into it.  If there were more than one view transform (e.g. several film stocks, ACES, or a digital camera rendering), then each display would exist in these flavors and one might have dcip3_film, dcip3_aces, dcip3_arri_k1s1, etc.  The output of each of these transforms is intended to be put on a DCI-P3 display, but there is no independent definition of simply what "DCIP3" actually means (i.e., just the EOTF).

This makes it difficult to integrate with the way the Mac and Windows OS and many applications (e.g. from Adobe) manage graphics monitors via ICC profiles.  A user may be using a wide-gamut monitor that is neither "sRGB" or "DCIP3".  An ICC profile is available for the monitor but there is no way to utilize it since there is no way to get from "sRGB" or "DCIP3" to the ICC profile connection space.  Although one may argue that the config should already have displays defined for whatever the artist has on their desk and so there is no need to use the ICC profile, this is only possible within the confines of a given VFX facility.  In general, there is a need to be able to define configs that work more generically (e.g. for all users of ACES, or all users of a given software application).

In addition, the lack of constructs in the config for fundamental color science concepts such as display-referred color spaces and view transforms (i.e. rendering transforms or OOTFs) make configs more difficult for people (other than their author) to understand.  Although there are ways of hacking OCIO today to create something similar to display color spaces and view transforms, each config author tends to find their own hack which may be different from someone else's.  

This lack of a standard "official" way of representing these concepts also presents a barrier to understanding for people learning color science.  We should remember that color science is often regarded as a difficult subject to understand and many people who get tasked with maintaining OCIO configs have not had any formal training in color science.  So while part of the problem this proposal attempts to address is "mechanical" within the OCIO code/API, arguably the more important part of it is "conceptual" to try and better align OCIO with the current standards and best practices for managing color.

**CONFIG PROPOSAL**

In the displays/views section of the config, it would be possible to define a view that is decoupled into a view transform and display color space.  Here is an example showing the traditional and proposed way.  Note that these may be used side-by-side, the traditional method would not be removed.
```
displays:
  DCIP3:
    - !<View> {name: Film, view_transform: Film, display_colorspace: DCIP3}
    - !<View> {name: Colorimetry, view_transform: Colorimetry, display_colorspace: DCIP3}
    - !<View> {name: Log, colorspace: lm10}
    - !<View> {name: Raw, colorspace: nc10}
  sRGB:
    - !<View> {name: Film, view_transform: Film, display_colorspace: sRGB}
    - !<View> {name: Colorimetry, view_transform: Colorimetry, display_colorspace: sRGB}
    - !<View> {name: Log, colorspace: lm10}
    - !<View> {name: Raw, colorspace: nc10}
```

A new section of the config would be defined for display-referred color spaces.  The entries would mostly use the existing ColorSpace syntax.  The exception is that rather than a from/to_reference they would use from/to_display_reference.  Here is an example:
```
display_colorspaces:
  - !<ColorSpace>
    name: DCIP3
    from_display_reference: !<GroupTransform>
      children:
        - !<MatrixTransform> {matrix: [ ... CIE XYZ to DCIP3 ... ]}
        - !<ExponentTransform> {value: [2.6, 2.6, 2.6, 1], direction: inverse}
  - !<ColorSpace>
    name: sRGB
    from_display_reference: !<FileTransform> {src: CIE_XYZ_to_sRGB.clf}
```
Note that it is allowed (and one might even say encouraged) to define display color spaces using the same name as displays in the displays/views section of the config.  This is essentially defining the EOTF associated with that display.

A new section of the config would be defined for view transforms.  In terms of syntax and purpose, the new section is somewhat similar to the existing looks section in the sense that it is essentially defining transforms that may be referred to elsewhere in the config.  (Although both functionally and conceptually, of course, a view transform is quite different from a look transform.)  Here is an example:
```
view_transforms:
  - !<ViewTransform>
    name: Film
    from_reference: !<FileTransform> {src: ...}

  - !<ViewTransform>
    name: Colorimetry
    from_display_reference: !<FileTransform> {src: ...}
```
Note that it is possible for a view transform to convert either from the scene-referred reference color space to the display-referred reference color space OR to simply be a transform that starts and ends in the display-referred reference space.  Hence either the from/to_reference or from/to_display_reference may be used for the transform.  (Note that it is one or the other, both are not allowed for a given view transform.)

![image](https://user-images.githubusercontent.com/43830961/73706737-359ae000-46c7-11ea-90e4-fb0e79aad9b3.png)
**Figure 1**  The set of display color spaces are connected in a way that is very similar to the traditional color spaces.  A set of view transforms connects the two sets of color spaces.

![image](https://user-images.githubusercontent.com/43830961/73706693-13a15d80-46c7-11ea-8fc1-6e75df5371de.png)
**Figure 2**  A more detailed illustration of the path from a traditional or display color space, through a view transform, to a display color space.

**API PROPOSAL**

The proposal is that a display color space would behave the same from a user's point of view as a traditional color space.  So the config methods such as getColorSpace, getColorSpaceByIndex, etc. would accept or return display color spaces too.  The implication is that they would appear in menus in applications.  But as with all color spaces, they may be placed in the inactive_colorspaces list to hide them from menus if desired.

The function ColorSpaceTransform:BuildColorSpaceOps would accept a display color space as either a source or destination.  This involves handling four scenarios for (src, dst):
1. (CS,CS)
2. (DCS,DCS)
3. (CS,DCS)
4. (DCS,CS)

Where CS is a traditional color space and DCS is a display color space.  Scenarios 1 and 2 proceed as before.  (The second one uses the display_reference rather than the main reference.)  Scenarios 3 and 4 need to invoke a view transform but the problem is that there typically will not be a single unique view transform.

One option would be to throw (and indeed OCIO v1 does throw in some cases, for example if the it would require inverting a Lut3D).  However the proposal is that we allow the function to succeed and therefore need to identify a default view transform to be used.  The proposal is that the first view transform in the list will be the default.

The BuildColorSpaceOps function is what is used by the Config:getProcessor methods, so they inherit the above behavior.

Similar to ColorSpaceTransform, we would enhance DisplayTransform:BuildDisplayOps.  This would allow specifying a specific view transform to use for scenarios 3 or 4 above.  Also, we propose exposing this via an additional Config:getProcessor that takes a CS, DCS, and view transform as arguments.

Minimal changes should be needed to the Config class display/view methods.  In other words, methods such as getDisplay and getView continue to return whatever is defined in that section of the config.  The only difference being that now that those views may have been defined using the new syntax (view_transform + display_colorspace).

**ALTERNATIVE DESIGNS CONSIDERED**

When I made this proposal during a recent OCIO v2 working group meeting, Mark suggested that much of the functionality could also be obtained by allowing color spaces to define transforms to any other color spaces.  In other words, instead of a ColorSpace having a from/to_reference attribute, it could supply a transform that goes to any other ColorSpace.  This is an interesting alternative to consider.

In this scenario, it would not be necessary to define display color spaces as a new section of the config.  Config authors could introduce what amounts to a display reference space as a "normal" color space and other color spaces could transform to that rather than to the main reference space.  Any additional reference spaces could also be added as needed.

By coincidence, the color management system we designed at Autodesk (SynColor) has a similar feature.  It's possible to define color space transforms between arbitrary pairs of color spaces.  However we found that it added a fair amount of complexity since conversion between a pair of color spaces is then essentially a "routing" or "path finding" problem from network/graph theory.  In some cases there may be more than one path.  

For example, if you think of color spaces as airports, going from Montreal to Los Angeles might stop in either Chicago or Denver.  What heuristics should be used to pick the "best" route?  Also, adding a new color space may change the route that gets selected so the results of a previous conversion between that pair is now different.  E.g., the Los Angeles and Montreal "color spaces" may not have changed but the conversion between them is now different because a new Las Vegas "color space" was added and the route now goes that way.

Similarly, as other color spaces in the config are modified, it may become impossible to find a route between a pair of color spaces that were previously connectable.  Imagine the possibility of being trapped in sRGB (the horror!).

As we have been porting SynColor features to OCIO, one of the side-effects of that is some added complexity.  One topic that we give a lot of thought to is whether the benefits of the new feature outweigh that added complexity.  We also consider whether there is a way of achieving a similar feature to what we have in SynColor but with less complexity.

My opinion is that the added complexity of allowing arbitrary color space to color space connections outweighs the benefits.  Basically the only time we needed to use that feature in SynColor was in order to implement a display-referred connection space.  And given that, the above proposal is a way of doing that which adds far less complexity to the OCIO code base.

With the proposal being made, fairly minimal changes are needed to the existing OCIO routing mechanism (e.g., as expressed in ColorSpaceTransform:BuildColorSpaceOps).  This is because there is a single reference space on both the scene-referred and display-referred sides.  With the alternate proposal, this becomes much, much more complicated.

The other drawback of the alternative approach is that the configs themselves would become more difficult to understand.  There would no longer be a well-defined path between color spaces.  Also, adding or modifying a color space would have the potential to change the result of a previous conversion between two completely different color spaces.

The other question to consider is whether the introduction of the display color space concept is a "bug" or a "feature."  If one considers this a "bug," then the alternative proposal becomes more interesting since it avoid the need to introduce this concept.  However, my point of view on this is that introducing explicit display color spaces and view transforms is a desirable feature.  These are fundamental concepts of modern color science and there should be a clear and well-defined way of expressing these concepts in a config.

In summary, while the alternative design is an interesting option to consider, it is _not_ the approach I'm advocating as a solution to the display-referred color space issue.

**Comment by brechtvl** [comment]
In Blender we have similar problems, where the design of OpenColorIO seems more aimed at VFX facilities than individual users that wouldn't create their own config.

I think it would be great if there was a document with recommendations on how to write a configuration, perhaps even more important (to us) than design changes like this.

As I understand it at least, everything in the proposal would be possible if the configuration follows conventions like:
* Display names sRGB, DCIP3, .. correspond to the standard color spaces with those names.
* For every display, there is a color space with the same name defining the EOTF for that display.
* All displays should support the same list of view transforms.

If it's a more explicit part of the design that's helpful. But what I personally really care about is for configurations to really follow the ideas of displays and view transforms, and even OCIO ACES config seems to misuse the display concept by having an "ACES" display.
https://github.com/imageworks/OpenColorIO-Configs/blob/master/aces_1.0.3/config.ocio

**Comment by doug-walker** [comment]
Hi Brecht,  I agree with you that there is a problem with configs misusing the display/view section of the config.  In the case of the ACES config, the authors were forced to adopt that approach due to limitations in the ways important apps supported (or rather did not properly support) displays and views in their UIs.

However, this proposal does not change the app-facing aspect of views/dispays, so it neither helps nor hurts in regard to your point.  What may help would be clearer instructions in the documentation for app developers about how to present displays and views in their user interface.

Your bullet points are actually not correct:
"- Display names sRGB, DCIP3, .. correspond to the standard color spaces with those names."
Yes, this could be done today, but you still have the problem that each of those must have the view transform embedded, which may or may not be invertible.  (Or may or may not be designable so that the optimizer may remove pair inverses.)

"- For every display, there is a color space with the same name defining the EOTF for that display."
Again, these cannot go from a display-referred reference space (which doesn't exist) to a display, so they are not just the inverse EOTF.  They are that plus a view transform.

"- All displays should support the same list of view transforms."
This is what some applications incorrectly assume.  This is not a requirement of OCIO and in fact does not make sense for some configs.  Applications should not make this assumption but that is a separate issue, not directly related to this proposal.

One could do as you propose and create these display color spaces with an identity view transform.  But then the config becomes a mess and applications would not know which color spaces to present.  For example, there might be an sRGB color space that goes directly from the scene-referred reference space to the display with simply an inverse EOTF.  This would create a lot of confusion for users since it is going to look bad (since it is missing the appropriate view transform).  So again, that is not a solution to the problem that we are trying to address.

Thanks for the feedback!

**Comment by KelSolaar** [comment]
@doug-walker: Sorry for not having commented earlier, being thinned from all the directions!

I like the proposal and it does make a lot of sense. I would be in favor of the alternative DAG design but I have not a strong opinion about it, it is mostly my nerd facet that thinks it would be fantastic but if pragmatism and experience have shown that it is not needed, well,  I'm equally happy with the simpler proposal.

Something that comes to mind though after re-reading it is whether there is not a case for explicit `scene_colorspace` and `from/to_scene_colorspace` keywords. It would maybe help to categorize things instead of having a generic `from/to` that is a mixed bag of many things and the new `from_display_reference/to_display_reference`.

**Comment by brechtvl** [comment]
@doug-walker, thanks for the reply, I don't want to take this off-topic. It just would be a shame to get this new functionality, but perhaps still struggle to get people to even adopt the display / view transform concept which I think is really a core feature of OCIO.

Anyway, from the point of view of an application developer the proposal by itself is great.

Just to clarify what I meant by the bullet points:

>  Yes, this could be done today, but you still have the problem that each of those must have the view transform embedded, which may or may not be invertible. (Or may or may not be designable so that the optimizer may remove pair inverses.)
> ..
>  Again, these cannot go from a display-referred reference space (which doesn't exist) to a display, so they are not just the inverse EOTF. They are that plus a view transform.

If the config is structured to do transforms similar to this proposal, then to me it seems OCIO can guarantee that the transforming between display spaces and the display-referred reference space is always possible.

Because the first part of the transform would be the same view transform that cancels out, and the second part would be an invertible EOTF? Anyway, I don't want to push this point too much since I understand it could be tricky for a config author to get right.

>  This is what some applications incorrectly assume. This is not a requirement of OCIO and in fact does not make sense for some configs. Applications should not make this assumption but that is a separate issue, not directly related to this proposal.

I understand there may be practical reasons why this is not possible, I wouldn't make it a hard requirement, we don't assume it either in Blender. For usability though, it's not great that when you switch to another display the view transform may no longer be available and the image looks entirely different.

**Comment by doug-walker** [comment]
@KelSolaar, I like your idea about scene_colorspace, although I will suggest to_scene_reference and from_scene_reference might be clearer.  I think we should allow these as synonyms for to_reference and from_reference.  In OCIOYaml we will save v2 configs using the new form.  I'll add this to the agenda for tomorrow's working group meeting.

@brechtvl, thanks for clarifying your points.

**Comment by KelSolaar** [comment]
> I think we should allow these as synonyms for to_reference and from_reference.

Yeah sorry, that what I was meant to write! Me failing at multi-processing  :)

**Comment by doug-walker** [comment]
Closing the issue as it has been implemented.  Created a new issue #1243 to capture some small refinements.
